File: 0055-jump-game/0055-jump-game.cpp
class Solution
{
    public:
        bool canJump(vector<int> &nums)
        {

            int n = nums.size();
            if(n == 1) return 1;
            int max_reach = nums[0];
            if (max_reach == 0) return 0;
            for (int i = 1; i < n; i++)
            {

                max_reach = max(max_reach - 1, nums[i]);
                if (max_reach == 0 && i != n - 1) return false;
            }
            return true;
        }
};

File: 0055-jump-game/0055-jump-game.java
class Solution {
    boolean solve(int[] nums, int ind, int target, int[] dp){
        if(ind == target) return true;
        if(ind > target) return false;
        
        if(dp[ind] == 1) return true;
        
        int maxreach = nums[ind];
        boolean fromNewInd = false;
        for(int jump = 1; jump <= maxreach; jump++){
            int newInd = ind + jump;
            fromNewInd = solve(nums, newInd, target, dp);
            if(fromNewInd) {
                dp[newInd] = 1;
                return true;
            }
        }
        dp[ind] = 0;
        return false;
    }
    
    public boolean canJump(int[] nums) {
        int n = nums.length;
        boolean[] dp = new boolean[100001];
        //Arrays.fill(dp, -1);
        //return solve(nums, 0, nums.length-1, dp);
        
        dp[n-1] = true;
        
        for(int i=n-2; i>=0; i--){
            int maxJump = Math.min(i + nums[i], n - 1);
            for(int j=i+1; j<=maxJump; j++){
                if(dp[j]) {dp[i] = true; break;}
            }
        }
        return dp[0];
    }
}